examples/event_loop/src/event_loop_example.cpp

```diff
@@ -0,0 +1,108 @@
+// Copyright 2023 Klepsydra Technologies AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <klepsydra/high_performance/event_loop_middleware_provider.h>
+#include <klepsydra/vision_ocv/file_image_stream_service.h>
+#include <klepsydra/vision_ocv/image_data_factory.h>
+#include <klepsydra/core/publisher.h>
+#include <klepsydra/core/subscriber.h>
+
+#include "spdlog/spdlog.h"
+int main(int argc, char *argv[]) {
+  if (argc < 2) {
+    spdlog::error("Input folder not provided.");
+    return -1;
+  } else if (argc < 3) {
+    spdlog::error("Output folder not provided.");
+    return -1;
+  }
+
+  const std::string inputFolderName = argv[1], outputFolderName = argv[2];
+  const int width = 500, height = 500;
+  const int totalCameras = 3, readTimes=6;
+
+  spdlog::set_level(spdlog::level::debug);
+  kpsr::vision_ocv::ImageDataFactory imageCreator(width, height, cv::IMREAD_REDUCED_GRAYSCALE_2, "camera_link");
+  kpsr::high_performance::EventLoopMiddlewareProvider<16> eventLoopProvider(nullptr);
+  std::vector<kpsr::Subscriber<kpsr::vision_ocv::ImageData>*> subscribers;
+  std::vector<kpsr::Publisher<kpsr::vision_ocv::ImageData>*> publishers;
+  std::vector<std::unique_ptr<kpsr::vision_ocv::FileImageStreamingService>> readFromDiskServices;
+  std::vector<std::thread> threads;
+  std::array<int, totalCameras> frameCounters{};
+
+  eventLoopProvider.start();
+  
+  std::string prefixChannel = "channel_";
+
+  // subscriber creation and listener registration.
+  for (int i = 0; i < totalCameras; i++) {
+    std::string channelName = prefixChannel+std::to_string(i);
+    subscribers.push_back(eventLoopProvider.getSubscriber<kpsr::vision_ocv::ImageData>(channelName));
+    subscribers[i]->registerListener(channelName, [i, &frameCounters, outputFolderName](const kpsr::vision_ocv::ImageData &event) {
+            std::string imgName = outputFolderName + "/camera_id_"+std::to_string(i)+"_frame" 
+              + std::to_string(frameCounters[i]++) + ".jpg"; 
```

Comment:
move all the constant parts (which do not depend on the event) out of the listener.  So the lambda function gets the result of `outputFolderName + "/camera_id_" + std::to_string(i) + "_frame"` and only does the `std::to_string(frameCounter[i]++) + ".jpg` inside.